### PR TITLE
Feat/azure devops backend

### LINKED ISSUE #1181: Microsoft Team Services / Azure as a backend?
Is there in the roadmap also support for team services (https://www.visualstudio.com/team-services/)? It is based on git and free for small companies and startups (private repositories), so very attractive for me.

### PR DESCRIPTION
**Pull request by benhulan**
Fixes #1181

**Summary**
This builds directly on top of some great work done in [this previous PR](https://github.com/netlify/netlify-cms/pull/3318) and includes this summary by @nathankitchen below.

Azure DevOps (formerly Visual Studio Online) is a popular platform for hosting Git repositories. This PR includes an additional backend to enable Netlify-CMS to work with Azure DevOps repositories. It interacts with the [Azure DevOps Git API](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/?view=azure-devops-rest-5.1) using an implicit-auth token from Azure AD.

Backend support includes:
* Media file support for publicly inaccessible files
* Editorial workflow is now fully functional

The following user stories have been tested against our own Azure repos and are found to be working:
 "As a Netlify-cms user with Azure repos..."
 - "I can log in and view my content"
 - "I can create, get, update and delete media files"
 - "I can create, get, update and delete structured content in markdown and JSON files"
 - "PRs created without the editorial workflow can be managed with the editorial workflow"

A few TypeScript tests are failing but this feature now works and I would love some help getting it across the finish line
 
**Test plan**
Some tests have been added. At one point, @erezrokah offered to add more

Additional thanks goes out to @jburich, @MKrupauskas, @chrismade and @scott-armiger

Non-mandatory cute animal:
Blue Heeler Puppy

### CODE REVIEW

**Comment by benhulan** [comment]
Note to self:  Update README.md to require `backend: preview_context: '-preview'` in config.yml to work with editorial workflow

**Comment by benhulan** [comment]
Hmm. I have "Allow edits from maintainers" checked. I could invite you as a collaborator on my branch copy

**Comment by erezrokah** [comment]
> Hmm. I have "Allow edits from maintainers" checked. I could invite you as a collaborator on my branch copy

So it's probably something with my set up. Let me check that.

**Comment by benhulan** [comment]
Thanks for your quick work, @erezrokah! I removed the tests I copied from the GitHub backend. Ideally we would re-write these at some point in the near future but I am out of time. Also updated the README with more recent info

**Comment by benhulan** [comment]
Rebased over current master branch

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 1]
```diff
@@ -0,0 +1,1007 @@
...
+  }
+
+  uploadAndCommit(
+    items: AzureCommitItem[],
+    comment = 'Creating new files',
+    branch: string = this.branch,
+    newBranch = false,
+  ) {
+    return this.getRef(branch).then(async (ref: AzureRef) => {
+      if (ref == null) {
```
Comment (erezrokah) [inline packages/netlify-cms-backend-azure/src/API.ts:590]: Shouldn't this be an error? If `branch` was passes we want to make sure to commit to it and not to the default one.
If `branch` was not passes it defaults to the default one which in that case calling `getRef` again with the same value is redundant.

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 2]
```diff
@@ -0,0 +1,1007 @@
...
+    await this.requestJSON({
+      method: 'POST',
+      url: `${this.endpointUrl}/refs`,
+      body: JSON.stringify(deleteBranchPayload),
+    });
+  }
+
+  uploadAndCommit(
+    items: AzureCommitItem[],
+    comment = 'Creating new files',
```
Comment (erezrokah) [inline packages/netlify-cms-backend-azure/src/API.ts:585]: When is the default value for `comment` is used?

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 3]
```diff
@@ -0,0 +1,1007 @@
...
+    const diffsWithIds = await Promise.all(
+      diffs.map(async d => {
+        const path = d.item.path;
+        const newFile = d.changeType === 'add';
+        const id = d.item.objectId;
+        return { id, path, newFile };
+      }),
+    );
+    const label = pullRequest.labels.find(l => isCMSLabel(l.name, this.cmsLabelPrefix));
+    const labelName = label && label.name ? label.name : DEFAULT_NETLIFY_CMS_LABEL_PREFIX;
```
Comment (erezrokah) [inline packages/netlify-cms-backend-azure/src/API.ts:663]: We should only be reading pull requests with a CMS label attached to them.
Not sure when `label` can be undefined, and in any case we shouldn't default to `DEFAULT_NETLIFY_CMS_LABEL_PREFIX`

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 4]
```diff
@@ -0,0 +1,1007 @@
...
+    const status = labelToStatus(labelName, this.cmsLabelPrefix);
+    const updatedAt = pullRequest.closedDate;
```
Comment (erezrokah) [inline packages/netlify-cms-backend-azure/src/API.ts:665]: `closedDate` doesn't look like the correct value to use here:
https://docs.microsoft.com/en-us/rest/api/azure/devops/git/pull%20requests/get%20pull%20request?view=azure-devops-rest-6.0#gitpullrequest

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 5]
```diff
@@ -0,0 +1,1007 @@
...
+    return {
+      collection,
+      slug,
+      status,
+      diffs: diffsWithIds,
+      updatedAt,
+    };
+  }
+
+  isUnpublishedEntryModification(path: string, branch: string) {
```
Comment (erezrokah) [inline packages/netlify-cms-backend-azure/src/API.ts:675]: Where is this method used?

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 6]
```diff
@@ -0,0 +1,1007 @@
...
+
+    return this.requestJSON({
+      url: `${this.endpointUrl}/pushes`,
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json; charset=utf-8' },
+      body: JSON.stringify(push),
+    });
+  }
+
+  contentKeyFromBranch(branch: string) {
```
Comment (erezrokah) [inline packages/netlify-cms-backend-azure/src/API.ts:758]: You can re-use these methods from `netlify-cms-lib-util/src/APIUtils.ts`

File: packages/netlify-cms-ui-default/src/AuthenticationPage.js [hunk 7]
```diff
@@ -95,7 +95,7 @@ AuthenticationPage.propTypes = {
   loginErrorMessage: PropTypes.node,
   renderButtonContent: PropTypes.func,
   renderPageContent: PropTypes.func,
-  t: PropTypes.func.isRequired,
+  t: PropTypes.func,
```
Comment (erezrokah) [inline packages/netlify-cms-ui-default/src/AuthenticationPage.js:98]: `t` is actually required

File: scripts/webpack.js [hunk 8]
```diff
@@ -20,13 +20,11 @@ const rules = () => ({
...
     },
   }),
-  css: () => [
-    {
-      test: /\.css$/,
-      include: ['ol', 'redux-notifications', 'react-datetime', 'codemirror'].map(moduleNameToPath),
-      use: ['to-string-loader', 'css-loader'],
-    },
-  ],
+  css: () => ({
```
Comment (erezrokah) [inline scripts/webpack.js:23]: Any reason for this change?

File: yarn.lock [hunk 9]
```diff
@@ -11553,6 +11553,11 @@ jquery@^3.4.0, jquery@^3.4.1:
   resolved "https://registry.yarnpkg.com/jquery/-/jquery-3.5.1.tgz#d7b4d08e1bfdb86ad2f1a3d039ea17304717abb5"
   integrity sha512-XwIBPqcMn57FxfT+Go5pzySnm4KWkT1Tv7gjrpT1srtf8Weynl6R273VJ5GjkRb51IzMp5nbaPjJXMWeju2MKg==
 
+js-base64@^2.5.1:
```
Comment (erezrokah) [inline yarn.lock:11556]: I think this is redundant

File: packages/netlify-cms-lib-util/src/APIUtils.ts [hunk 10]
```diff
@@ -2,7 +2,7 @@ export const CMS_BRANCH_PREFIX = 'cms';
 export const DEFAULT_PR_BODY = 'Automatically generated by Netlify CMS';
 export const MERGE_COMMIT_MESSAGE = 'Automatically generated. Merged on Netlify CMS.';
 
-const DEFAULT_NETLIFY_CMS_LABEL_PREFIX = 'netlify-cms/';
+export const DEFAULT_NETLIFY_CMS_LABEL_PREFIX = 'netlify-cms/';
```
Comment (erezrokah) [inline packages/netlify-cms-lib-util/src/APIUtils.ts:5]: Don't think we need to export this `const`

File: packages/netlify-cms-backend-bitbucket/src/API.ts [hunk 11]
```diff
@@ -653,8 +653,8 @@ export default class API {
       params: {
         pagelen: 50,
         q: oneLine`
-        source.repository.full_name = "${this.repo}" 
-        AND state = "${BitBucketPullRequestState.OPEN}" 
+        source.repository.full_name = "${this.repo}"
```
Comment (erezrokah) [inline packages/netlify-cms-backend-bitbucket/src/API.ts:656]: Do we need this change?

File: packages/netlify-cms-backend-azure/package.json [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  ],
+  "sideEffects": false,
+  "scripts": {
+    "develop": "yarn build:esm --watch",
+    "build": "cross-env NODE_ENV=production webpack",
+    "build:esm": "cross-env NODE_ENV=esm babel src --out-dir dist/esm --ignore **/__tests__ --root-mode upward --extensions \".js,.jsx,.ts,.tsx\""
+  },
+  "dependencies": {
+    "common-tags": "^1.8.0",
+    "js-base64": "^2.5.1",
```
Comment (erezrokah) [inline packages/netlify-cms-backend-azure/package.json:25]: I think we're using `v3` of `js-base64`

**Review by erezrokah — CHANGES_REQUESTED** [review]
Thank you for picking this up @benhulan. I know some people have been waiting for this for quite some time.
Haven't finished the review yet, but wanted to share what I have so far.
Please let me know what you think.

I haven't tested the functionality yet, just went over some parts of the code.

As for me not being able to push to the branch, I think it's related to it being a fork of a fork...

**Comment by benhulan** [comment]
Thanks for your feedback. Most of your comments relate to content and code written by other developers. I noticed the editorial workflow is broken again after rebasing over the latest master, so I'm going to rebase again and try to fix it, and will address as many of these comments as I can.

File: packages/netlify-cms-lib-util/src/APIUtils.ts (hunk 10 above)
Comment (benhulan) [reply packages/netlify-cms-lib-util/src/APIUtils.ts:5]: I'm using this in the implementation file but if you prefer not exporting it, I can simply hard-code the string in the other file.

Comment (erezrokah) [reply packages/netlify-cms-lib-util/src/APIUtils.ts:5]: See my other comments. I think once they are resolved we won't need to export it

File: packages/netlify-cms-backend-bitbucket/src/API.ts (hunk 11 above)
Comment (benhulan) [reply packages/netlify-cms-backend-bitbucket/src/API.ts:656]: Nope

File: yarn.lock (hunk 9 above)
Comment (benhulan) [reply yarn.lock:11556]: Good catch

File: packages/netlify-cms-backend-azure/package.json (hunk 12 above)
Comment (benhulan) [reply packages/netlify-cms-backend-azure/package.json:25]: Ran yarn upgrade and it's gone. Hope to push again before EOD

File: dev-test/config.yml [hunk 13]
```diff
@@ -245,3 +245,14 @@ collections: # A list of collections the CMS should be able to edit
               - { label: 'Date', name: 'date', widget: 'date' }
               - { label: 'Image', name: 'image', widget: 'image' }
               - { label: 'File', name: 'file', widget: 'file' }
+  - name: pages # a nested collection
```
[the remark this answers is not in the document]
Comment (benhulan) [reply dev-test/config.yml:248]: You mean just the comment `# a nested collection` right? Not the whole pages collection?

[the remark this answers is not in the document]
Comment (erezrokah) [reply dev-test/config.yml:248]: The whole pages collection (no changes should be made to this file)

File: packages/netlify-cms-ui-default/src/AuthenticationPage.js (hunk 7 above)
Comment (benhulan) [reply packages/netlify-cms-ui-default/src/AuthenticationPage.js:98]: Requiring this throws an error during login with Azure.

File: scripts/webpack.js (hunk 8 above)
Comment (benhulan) [reply scripts/webpack.js:23]: Not that I know of; I assumed this was done by the linter or `yarn format`

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 14]
```diff
@@ -0,0 +1,1007 @@
...
+
+export const API_NAME = 'Azure DevOps';
+
+export class AzureRepo {
+  org: string;
+  project: string;
+  name: string;
+
+  constructor(location?: string | null) {
+    if (!location || !/^[^/]+\/[^/]+(|\/[^/]+)$/gi.test(location)) {
```
[the remark this answers is not in the document]
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts:34]: I'm not sure what array you're referring to. Maybe that link doesn't go where you intend. One quick and dirty fix might be: `if (!location || location.indexOf('/') !== -1) { ...`

**Comment by chrismade** [comment]
I was the one who added the "why to invest" into the readme - at a time when I when my limited expertise were doing more harm than good to the code - now I think we found the right people and hence we could remove this part resp move it to a more appropriate place in the netlify-cms docs - happy to see so much progress now and hope I can contribute then to the docs e.g. use azure dev op pipeline to deploy to blob storage and flush the CDN - and yes, true, github now has free private repos *now*

**Comment by benhulan** [comment]
I'm about to push up my changes which should fix most of these comments. I did not attempt to convert classes, so hopefully that's not a blocker. @erezrokah when you push an update to this branch I see it on my personal repo and can incorporate the changes and push them back here if that helps

File: dev-test/config.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (benhulan) [reply dev-test/config.yml]: Ah. I misunderstood. Pushed some changes and requested re-review but I'll remove this and push again now

File: packages/netlify-cms-ui-default/src/AuthenticationPage.js (hunk 7 above)
Comment (erezrokah) [reply packages/netlify-cms-ui-default/src/AuthenticationPage.js:98]: We should pass the `t` function to the component to fix the error. I think it should be fixed here:
https://github.com/netlify/netlify-cms/pull/4427/files#diff-15ef45b7f567d47fc75af46ae70e53100f7ff3a3fa2b61a07809c8c4fed34e37R63

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 15]
```diff
@@ -0,0 +1,1007 @@
...
+      {
+        'Content-Type': 'application/json; charset=utf-8',
+        Origin: '*',
+      },
+      req,
+    );
+
+    req = unsentRequest.withParams(
+      {
+        'api-version': this.apiVersion + '-preview',
```
[the remark this answers is not in the document]
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts:360]: The '-preview' suffix allows editorial workflow to work

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 16]
```diff
@@ -0,0 +1,1007 @@
...
+        );
+      }
+
+      const processedAzureTreeEntries: AzureGitTreeEntryRef[] = [];
+
+      for (const f of azureTreeEntries) {
+        f.relativePath = `${path}/${f.relativePath}`;
+        let entry: AzureGitTreeEntryRef | undefined = f;
+        // If AzureGitTreeEntryRef is still a tree object, we need to drill further to get to the blob
+        if (f.gitObjectType === 'tree') {
```
[the remark this answers is not in the document]
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts:523]: This is used for both 'files' and 'folders' collection types. I believe the files collection type returns a tree before we can access the blob

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 17]
```diff
@@ -0,0 +1,1007 @@
...
+  repo: AzureRepo;
+  endpointUrl: string;
+  initialWorkflowStatus: string;
+  commitAuthor?: AzureCommitAuthor;
+  cmsLabelPrefix: string;
+
+  constructor(config: AzureApiConfig, token: string) {
+    this.repo = config.repo;
+    this.apiRoot = trim(config.apiRoot, '/') || 'https://dev.azure.com';
+    this.endpointUrl = `${this.apiRoot}/${this.repo?.org}/${this.repo?.project}/_apis/git/repositories/${this.repo?.name}`;
```
[the remark this answers is not in the document]
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts:337]: I've removed the `?` operator

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 18]
```diff
@@ -0,0 +1,1007 @@
...
+
+    return req;
+  };
+
+  buildRequest = (req: ApiRequest) =>
+    flow([
+      unsentRequest.withRoot(this.apiRoot),
+      this.withAuthorizationHeaders,
+      this.withAzureFeatures,
+      // unsentRequest.withTimestamp,
```
[the remark this answers is not in the document]
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts:373]: Pulling in `buildRequest` from the BitBucket API backend seems to work.

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 19]
```diff
@@ -0,0 +1,1007 @@
...
+   * Get the name of the current user by hitting the VS /me endpoint to
+   * return an AzureUser object.
+   */
+  user = (): Promise<AzureUser> => {
+    return this.requestJSON({
+      url: 'https://app.vssps.visualstudio.com/_apis/profile/profiles/me',
+    }) as Promise<AzureUser>;
+  };
+
+  async retrieveMetadata(contentKey: string) {
```
[the remark this answers is not in the document]
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts:406]: Pulling in `request` from the BitBucket API backend seems to work.

File: packages/netlify-cms-backend-azure/src/API.ts (hunk 4 above)
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts:662]: Azure doesn't seem to expose a generic `updatedAt` value.

File: packages/netlify-cms-backend-azure/src/API.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts:337]: ...but this makes TypeScript complain. `Config.repo` is optional in `netlify-cms-lib-util`

[the remark this answers is not in the document]
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts:339]: reinstated the `?`s for now

File: packages/netlify-cms-backend-bitbucket/src/API.ts (hunk 11 above)
Comment (benhulan) [reply packages/netlify-cms-backend-bitbucket/src/API.ts:656]: TBH I don't see anything changed here. It looks like these two lines were cut and then restored. Not sure why git thinks there's a change. See SHA `eb6b268`

**Comment by jballe** [comment]
Thank you for your effort on this (all involved). 
I would like to use this feature and Azure DevOps to allow use existing credentials in Azure AD. 

I have not really been involved with contributing to this repo before, can I assist with documentation, testing or similar? Is a preview build available anywhere or would I need to build it locally?

Again, thank you for your effort both this product available for the public and specifically this feature.

**Comment by erezrokah** [comment]
@benhulan I added 2 more commits to https://github.com/netlify/netlify-cms/commits/feat/azure-devops-backend as code cleanup.
I tried setting this up locally, but couldn't get pass authentication. I'm getting:
image

What am I missing?

**Comment by erezrokah** [comment]
Also it seems this PR doesn't handle pagination. What are the page limits in Azure?

**Comment by benhulan** [comment]
Thank you -- I merged your code and am trying to test on my end. 

I have not seen that error message before. I assume you generated that application ID with a Microsoft Account. When I log in with my `tenantId` and `appId`, I am redirected to my org's SSO login page, and then after logging in, I am redirected back to Netlify-CMS with no issues.

```
  identity_url: https://login.microsoftonline.com/{tenantId}
  app_id: {appId}
```

**Comment by benhulan** [comment]
> Also it seems this PR doesn't handle pagination (this means sorting, filtering, i18n, nested collections won't work properly). What are the page limits in Azure?

It looks like we may be able to pass `$top` and `continuationToken` to list Azure Git Refs.

https://docs.microsoft.com/en-us/rest/api/azure/devops/git/refs/list?view=azure-devops-rest-6.0

**Comment by erezrokah** [comment]
> I assume you generated that application ID with a Microsoft Account

I followed the instructions in `azure-backend.md`. Maybe some of the people watching this PR can help?

**Comment by benhulan** [comment]
> > I assume you generated that application ID with a Microsoft Account
> 
> I followed the instructions in `azure-backend.md`. Maybe some of the people watching this PR can help?

Unfortunately Microsoft's documentation is not very straightforward. We had much more extensive documentation before. I followed the CLI instructions and it worked, but I was asked to delete those because they were too confusing. From `a9929266139c48cf06b346ceae8e7f73cce91608`:


### Create an app in AAD (Azure Active Directory)

#### Manual version
  1. Sign into https://portal.azure.com
  2. Search for "Azure Active Directory" (also known as AAD)
      1. Double check that you see the tenant associated with your Azure DevOps instance!
      2. If not, use the directory switcher in the top bar to go to the right directory.
  3. In the Azure AD blade, find "App Registrations".
  4. Click "New registration"
  5. Enter a display name, for example: `Netlify CMS for Azure DevOps`
  6. Choose "Single tenant" mode, (multi-tenant hasn't been tested/here be dragons)
  7. Enter your redirect URL, e.g. http://localhost:8080
  8. Press "OK" to create the app
  9. Go to "API permissions"
      1. Click "Add a permission"
      2. Click "Azure DevOps"
      3. Under "Delegated permissions", check "user_impersonation"
      4. Click "Add permissions"
  10. To save users having to consent, hit "Grant admin consent for (your tenant)"
  11. Go to "Authentication"
      1. Under "Web", find the "Implicit grant" section
      2. Check "Access tokens" and "Identity tokens"
      3. Enter "https://dev.azure.com/(yourorgname) in the redirect URIs list (*Note to self: check this is necessary!*)
      4. Click "Save"
  12. Go to "Overview"
      1. Note the "Application (client) ID" for the app_id config.yml setting.
      2. Note the "Directory (tenant) ID" to append to the identity_url setting.

When the dev environment starts, hit "Login" and sign in with an Azure AD account that has access to the Azure DevOps repo!

#### Script version
*While it is possible to automate creation of the service principal as described below, it's usually easier to just do it in the portal. Also, the below scripts haven't been double checked to validate that they achieve the same as the instructions above*

In case you have 'Azure CLT' (Command Line Tools) installed you should have an 'az' command in the commandline:

```
az --version
```
will output a list of installed submodules with its version information so you know Azure Tools are installed correctly - 

please use your favorite search engine in case it isn't - where to download and how to install - and then how to prepare for its first use: 

```
az login

```
require to enter your Azure username and password - same that you use for https://portal.azure.com and https://dev.azure.com - both have forever free options as of this writing).

```
az ad sp create-for-rbac --name somename --password somepassword 
+Netlify CMS was recently converted from a single npm package to a "monorepo" of over 20 packages.
+That's over 20 Readme's! This is a work in progress.
 
```
+In the meantime, you can:
 
for example:
```
az ad sp create-for-rbac --name netlifycms001 --password smcyfilten 
 
```
please pay attention to this output - should look like this:
 
```
{ 
  "appId": "5d5d5d5d-eeee-4444-aaaa-ffffffffffff", 
  "displayName": "netlifycms001", 
  "name": "http://netlifycms001", 
  "password": "smcyfilten", 
  "tenant": "71717171-3333-4545-bbbb-999999999999" 
}  

```
here we find the appId and tenant_id we need for the config.yml - but hold on, we are no done yet

[read more here](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest#az_ad_sp_create_for_rbac)

### change the app settings to allow creation of tokens

take your appId from above and issue this command:


```
az ad app update --id=5d5d5d5d-eeee-4444-aaaa-ffffffffffff  --oauth2-allow-implicit-flow=true --reply-urls="http://localhost:8080"
```
there is no output expected

you may want to add more reply-urls when you deploy netlify-cms to a real website - that's no problem you can have a list here - or even remove localhost:8080 if this app is not used for development (which is recommended).

### allow API usage for dev.azure.com

**Comment by benhulan** [comment]
> I have not really been involved with contributing to this repo before, can I assist with documentation, testing or similar? Is a preview build available anywhere or would I need to build it locally?

@jballe We may still need to implement pagination but I'm hopeful we can get this across the finish line soon. Meanwhile you can copy my fork of this repo and checkout the `feat/azure-devops-backend` branch and run locally. This is working for me the vast majority of use cases. Presumably we will get pagination errors if we have more than 100 refs in a PR.

**Comment by benhulan** [comment]
> Also it seems this PR doesn't handle pagination (this means sorting, filtering, i18n, nested collections won't work properly). What are the page limits in Azure?

It appears this code is only getting one `AzureRef` at a time, for better or worse. I have built some nested collections and sorting works for me. I'm not sure how to test filtering and i18n.

File: packages/netlify-cms-backend-azure/src/API.ts (hunk 4 above)
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts]: I have swapped to the PullRequest `creationDate`. This still may not be the proper value, but at least it returns a useful value in the UI.

File: packages/netlify-cms-backend-azure/src/API.ts (hunk 2 above)
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts]: Good catch. There was a disconnect and we were trying to assign the value to "message", which did not work. Should be clear now.

File: packages/netlify-cms-backend-azure/src/API.ts [hunk 20]
```diff
@@ -0,0 +1,1007 @@
...
+  EditorialWorkflowError,
+  statusToLabel,
+  PreviewState,
+  readFileMetadata,
+} from 'netlify-cms-lib-util';
+import { DataFile } from 'netlify-cms-lib-util/src/implementation';
+
+export const API_NAME = 'Azure DevOps';
+
+export class AzureRepo {
```
[the remark this answers is not in the document]
Comment (benhulan) [reply packages/netlify-cms-backend-azure/src/API.ts]: Thanks for fixing

**Review by erezrokah — COMMENTED** [review]
@benhulan do you mind doing a round of testing after my changes?
I think I got everything covered but want to be sure

**Comment by benhulan** [comment]
@erezrokah I have tested and everything still seems to be functioning properly! The only thing we might consider is that if the user wants to include `api_version` in their `config.yml`, they need to include `-preview` in order for the editorial workflow to function properly. Something like:

`~/website/content/docs/azure-backend.md`

```yaml
backend:
  name: azure
  repo: organization/project/repo # replace with actual path
  tenant_id: tenantId # replace with your tenantId
  app_id: appId # replace with your appId
  api_version: '6.1-preview' # optional, but must include "-preview" for editorial workflow to function
```

**Comment by erezrokah** [comment]
> I have tested and everything still seems to be functioning properly! The only thing we might consider is that if the user wants to include `api_version` in their `config.yml`, they need to include `-preview` in order for the editorial workflow to function properly. Something like:

I actually didn't document that on purpose. We can only guarantee the functionally to work (not just editorial workflow) with the default API version set by the CMS.
I don't want to encourage users to change the default unless the have a very good reason to do so (e.g. debugging purposes).

**Pull request merged**